### DISCUSSION TITLE
Computing Jacobians with ragged arrays (Awkward Array)

### DISCUSSION THREAD

**Discussion by agoose77** [jax-ml/jax#12664] [Q&A]
Hi all!

I'm a developer over at the https://github.com/scikit-hep/awkward-1.0/ project. 

From our docs
> Awkward Array is a library for nested, variable-sized data, including arbitrary-length lists, records, mixed types, and missing data, using NumPy-like idioms.

I recently started looking at our prototype JAX integration, and I'm looking to add the ability to compute the Jacobian of a function, e.g.

```python
def func(array):
    return ak.concatenate([array, array])
```

Awkward Arrays are built by composing layout nodes, so a NumPy array with shape `(4, 3, 2)` is built as:
```xml
<RegularArray size='3'>
    <content><RegularArray size='2' >
        <content><NumpyArray dtype='int64' len='24'>
            [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21
             22 23]
        </NumpyArray></content>
    </RegularArray></content>
</RegularArray>
```

Our initial approach to integration with JAX was to represent Awkward Arrays as singular atoms with a number of leaves, i.e. the example array above would be flattened into 
```python
([array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])], <AUX DATA>)
```

This means that both the input and output arrays of e.g. `func` are flattened in JAX terms into flat single buffers (in this case) alongside some auxillary data.

However, to make it easier to reason about, I've mocked up [a demo](https://gist.github.com/agoose77/04ee1325c3e67b8774dd80abfe55903c) where our entire layout nodes can be converted into trees (not just the root node). In this demo, there is only `NumpyArray` and `RegularArray` support, even though we have a wider range of node types.

In the example, the array is not ragged, so the input array has shape `(4, 3, 2)`, and the output of `func(array)` should have shape `(8, 3, 2)`. Therefore, the Jacobian should have shape `(8, 3, 2, 4, 3, 2)`. Using our layout system, we'd expect the following structure:
```xml
<RegularArray size='3'>                           # (8, 3, _, _, _, _) 
    <content><RegularArray size='2'>              # (_, 3, 2, _, _, _)
        <content><RegularArray size='4'>          # (_, _, 2, 4, _, _)
            <content><RegularArray size='3'>      # (_, _, _, 4, 3, _)
                <content><RegularArray size='2'>  # (_, _, _, _, 3, 2)
                    <content><NumpyArray dtype='float32'>...</NumpyArray></content> 
                </RegularArray></content>
            </RegularArray></content>
        </RegularArray></content>

    </RegularArray></content>
</RegularArray>
```

In the comments, a helpful key shows how the size (RHS and length LHS) are computed from each `RegularArray` node.

What I'd like from JAX is to ask our tree hooks to:
1. unflatten the argument node (with actual shape `(4,3,2)`) with the flat `8*3*2*4*3*2=1152` Jacobian values
2. unflatten the result node with (1)

I have played around with this a bit already, but it's clear to me that I'm not 100% confident on the details of what JAX is doing. Rather than guessing, I was hoping that I might get some pointers from the JAX team as to how to proceed from here.

Please feel free to follow up questions if anything needs explaining further. Thanks!

**Comment by jakevdp** [comment]
Hi - happy to help with this, but I'm not sure how to interpret your XML-like syntax for array representation. What do the fields mean? How should I interpret `(8, 3, _, _, _, _)` on an outer field? Is this something like a jagged array with five rows where the first row is size 8 and the second row is size 3? What do the underscores mean? And how should I understand the difference between a `RegularArray` and a `NumpyArray`?

(Sorry if this is standard stuff... if you have docs about these representations you can point to, I may be able to read up and better understand what you're asking)

**Reply by agoose77** [reply]
Thanks @jakevdp, just improving [the example](https://gist.github.com/agoose77/04ee1325c3e67b8774dd80abfe55903c) to make it more comprehensible.

>  Is this something like a jagged array with five rows where the first row is size 8 and the second row is size 3?

No, in this example everything is regular. The key just shows how parts of the array "shape" are built from individual layout nodes: the values in parentheses are the `shape` of the array (though shapes break down with ragged arrays, they're fine here because everything is regular).

Our `RegularArray` nodes operate upon an array-like node (e.g. NumpyArray, or RegularArray), and give it an extra dimension. The `size` parameter determines the inner dimension, and the length of the node then follows from `len(content) // self.size`. Therefore, in the key, `(8, 3, _, _, _, _)` describes a RegularArray node that has an inner size of 3, and a corresponding length of 8.

So, a `NumpyArray` is a thin wrapper around a Numpy array. It has a shape, etc. `RegularArray` is a single layout node that adds _one_ extra (regular) dimension. Usually in Awkward Array we have 1D `NumpyArray`s and many `RegularArray`s, but for compatibility with NumPy, we like to support `len(shape)>1` arrays too. Although `RegularArray` might look redundant - why not just use `NumpyArray`s with an extra dimension? - they're useful in mixing regular and ragged dimensions, e.g. `10 * 4 * var * 8`.

There is a small section in our user guide (WIP) that talks about our layout nodes: https://awkward-array.org/how-to-create-constructors.html#content-regulararray

**Reply by agoose77** [reply]
To clarify, the part that I am uncertain about is how JAX wants to build the result using the pytrees API. Right now, it _seems_ like Jax is unflattening the deepest leaf of the _result_ array with the Jacobian matrix (i.e. the `NumpyArray` with shape `(48,)` becomes an array with shape `(48, 24)`).

**Reply by jakevdp** [reply]
Yeah, I see. So the way to think about PyTrees is that internally JAX sees everything as a tuple of arrays. So if you take the jacobian with respect to a pytree structure, it will (1) flatten the structure into a tuple of arrays (2) compute the jacobian with respect to each array, and (3) unflatten the tuple back into the pytree structure. Does that make sense?

**Reply by jakevdp** [reply]
To make this kind of data structure play well with batching transforms (note that jacobian is basically just a composition of autodiff and batching) we've found it useful to design data structures so that this kind of batching of internal nodes is understood by the schema: an example of this is in `jax.experimental.sparse.BCOO`, which is a generalization of a standard `COO` sparse data structure in which the data and indices buffers can be batched, and this is logically equivalent to adding a leading batch dimension to the sparse array. This makes BCOO immediately compatible with JAX `vmap`, `jacobian`, and other batch-based JAX transforms.

I haven't thought much about jagged arrays, but I wonder if you couldn't take a similar approach here, and define a data structure built around support for batching of internal buffers?

**Reply by agoose77** [reply]
>  Does that make sense?

Yes, I'm comfortable with the general gist of what's going on. My first question is whether this> Does that make sense?

Yes, I'm comfortable with the basics of what JAX is doing here. In my case, I have a single array for the input and another for the output, so JAX is giving me a 2D array. I *think* the issue I'm running into is that JAX assumes I want the Jacobian in the form that JAX gives me. In Awkward Array, the Jacobian itself will be regular (the JAX representation of an Awkward Array is always a list of raveled buffers), but its form in the final Awkward Array might not be. I feel like JAX assumes that I can directly consume the Jacobian and store it in the pytree. Am I explaining clearly enough?

In particular, does JAX rely on the user to store the computed Jacobian matrix such that it can be retrieved as a 2D array later by `tree_flatten`? My hunch is yes, and that it is later used by the batching. If so, perhaps we need to implement our own Jacobian function using Jax that accounts for the fact that our arrays are ragged.

**Reply by agoose77** [reply]
BCOO looks interesting. Are you referring to [these lines](https://github.com/google/jax/blob/9d12e216d52c5deee46fc50435d4e8dd45f66b79/jax/experimental/sparse/bcoo.py#L2438-L2455) that appear to implement vmap support?

Also, where does the BCOO class implement a `tree_unflatten`? I can't see one in the source or base class for `BCOO`.

**Reply by jakevdp** [reply]
I think your options are:

1. Change how you represent ragged arrays so that the representation is closed under JAX's native definition of batching (i.e. adding a batch dimension to each buffer in the flattened tree representation results in a valid ragged array data structure). This is probably fairly heavy-weight, because it would involve generalizing the data structure that you use to represent raggedness. But the advantage is that the result would just work with JAX's native transforms
2. Define your own version of the `jacobian` transform, the `jacrev` transform, the `jacfwd` transform, the `vmap` transform, and every other transform that involves batching. This is probably the worst approach because it means your objects won't actually be compatible with JAX; rather they'll be compatible with your own JAX-like API that you have to maintain.
3. Use the experimental `vmappable` API to override the way that batching transforms interact with your objects, so that `jax.vmap` (and by extension `jax.jacobian`, etc.) will return the expected awkward array data structure. This hasn't been used much, but there is one example [here](https://github.com/google/jax/blob/9d12e216d52c5deee46fc50435d4e8dd45f66b79/jax/experimental/sparse/bcoo.py#L2438-L2455) for the sparse BCOO case. I'm less confident this will work, because we haven't really kicked the tires much on the vmappable overrides. Also there's no real documentation of this feature, so it would require some hacking to figure out what's possible.

If it were me, I'd probably go for approach 1 (that's served us really well for BCOO), but I understand if you don't want to extensively modify the way you represent ragged arrays.

**Reply by jakevdp** [reply]
`tree_flatten` and `tree_unflatten` for BCOO are defined here:
https://github.com/google/jax/blob/9d12e216d52c5deee46fc50435d4e8dd45f66b79/jax/experimental/sparse/_base.py#L59-L65
https://github.com/google/jax/blob/9d12e216d52c5deee46fc50435d4e8dd45f66b79/jax/experimental/sparse/bcoo.py#L2434-L2435

**Reply by agoose77** [reply]
Oh thanks. I misread the `tree_unflatten` implementation as abstract in the `_base` module.

**Reply by jakevdp** [reply]
I think either approach (1) or (3) could be really interesting in this case... let me know if you want to do a video chat to brainstorm / pair-code. It could be fun 😀

**Reply by agoose77** [reply]
Thanks Jake. I think the final two options look like those that I was expecting. I've been poking around in the JAX source, and as I understand it, the reason that I'm seeing square 2D arrays in the `unflatten` routines is because JAX is doing the equivalent of `jnp.eye(len(y))`, where `y` is the output of the function i.e. [in the cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#the-implementation-of-jacfwd-and-jacrev). Then, the VJP which maps from the codomain to the domain is batched using `vmap`, which produces an array of `(len(y), len(x))`.

In your suggestion to change the representation, I am not confident that such a thing would be possible, at least as I understand it. Awkward Array is built around composition of layouts, and this system is fairly well established. 

That said, I know that I do not have a strong grasp yet on what I expect JAX to do here. In particular, what does it mean for JAX to unflatten an existing node with new data of an additional dimension. If I can understand this part better, then it might be feasible to map from JAXs representation into our interpretation, in such a way that we can reverse the interpretation in `tree_flatten` when JAX needs to read the flattened tree again.

**Reply by agoose77** [reply]
@jakevdp hmm, GitHub didn't update with your most recent response. We're trying to get a major release out in the next couple of months, so we don't have a huge amount of time to work on this right now. However, I can tell that it would be very productive (for me) to get a sense of how to think about this, and your offer is most appreciated. Are you around next week some time before Friday? That might be a bit presumptuous...!

**Reply by jakevdp** [reply]
Sounds good - I'll send you an email at the address listed on your github profile
